src/module_3/push_mvp.py

```diff
@@ -0,0 +1,375 @@
+    and save them in a single figure.
+
+    Parameters:
+    - model_name (str): Name of the model for labeling the curves.
+    - y_pred (array-like): Predicted probabilities or scores.
+    - y_test (array-like): True labels.
+    - save_dir (str, optional): Directory to save the generated figure.
+    If None, the figure will not be saved.
+
+    Returns:
+    - None
+    """
+
+    # Create a timestamp for unique filenames
+    timestamp = datetime.now().strftime("%Y%m%d%H%M%S")
+
+    # ROC Curve
+    fpr, tpr, _ = roc_curve(y_test, y_pred)
+    roc_auc = auc(fpr, tpr)
+
+    # Precision-Recall Curve
+    precision, recall, _ = precision_recall_curve(y_test, y_pred)
+    pr_auc = auc(recall, precision)
+
+    plt.figure(figsize=(10, 5))
+
+    plt.subplot(1, 2, 1)
+    plt.plot(fpr, tpr, label=f"ROC (AUC = {roc_auc:.2f}) - {model_name} (C = {C:.2e})")
+    plt.plot([0, 1], [0, 1], color="navy", linestyle="--")
+    plt.xlim([0.0, 1.0])
+    plt.ylim([0.0, 1.05])
+    plt.xlabel("False Positive Rate")
+    plt.ylabel("True Positive Rate")
+    plt.title("ROC")
+    plt.legend(loc="lower right")
+
+    plt.subplot(1, 2, 2)
+    plt.plot(
+        recall, precision, label=f"PR (AUC = {pr_auc:.2f}) - {model_name} (C = {C:.2e})"
+    )
+    plt.xlim([0.0, 1.0])
+    plt.ylim([0.0, 1.05])
+    plt.xlabel("Recall")
+    plt.ylabel("Precision")
+    plt.title("Precision-Recall Curve")
+    plt.legend(loc="lower left")
+
+    plt.tight_layout()
+
+    if save_curves_path:
+        # Define the filename with a timestamp
+        figure_filename = f"Evaluation_Curves_{model_name}_C={C}_{timestamp}.png"
+        figure_path = os.path.join(save_curves_path, figure_filename)
+
+        plt.savefig(figure_path)
+
+    plt.show()
+
+
+def train_evaluate_model(
+    X_train: pd.DataFrame,
+    y_train: pd.Series,
+    X_val: pd.DataFrame,
+    y_val: pd.Series,
+    save_model_path: str,
+):
+    """
+    Train, evaluate, and save a logistic regression model with hyperparameter
+    tuning for regularisation. Additionally, plot the curves for ROC and PR.
+
+    Parameters:
+    - X_train (pd.DataFrame): Training set features.
+    - y_train (pd.Series): Training set outcome.
+    - X_val (pd.DataFrame): Validation set features.
+    - y_val (pd.Series): Validation set outcome.
+    - save_model_path (str): Path to save the trained model.
+
+    Returns:
+    - None
+    """
+    # Define parameter grid for C values
+    # We know in advance that strong params works better
+    param_grid = {"lr__C": np.logspace(-8, -2, num=9)}
+
+    # Create the pipeline
+    pipeline = Pipeline(
+        [
+            ("scaler", StandardScaler()),
+            ("lr", LogisticRegression(penalty="l2", solver="liblinear")),
+        ]
+    )
+    try:
+        # Search for the best C hyperparameter
+        grid_search = GridSearchCV(
+            pipeline, param_grid, cv=5, scoring="roc_auc", n_jobs=-1
+        )
+        grid_search.fit(X_train, y_train)
```

Comment:
unlike the notebook, here you are finding the parameters of `StandardScaler` only in the train set which is the correct way of doing it